code/modules/movespeed/_movespeed_modifier.dm

```diff
@@ -0,0 +1,219 @@
+this spits out a final calculated value which is used as a modifer to last_move + modifier for calculating when a mob
+can next move
+
+Key procs
+* [add_movespeed_modifier](mob.html#proc/add_movespeed_modifier)
+* [remove_movespeed_modifier](mob.html#proc/remove_movespeed_modifier)
+* [has_movespeed_modifier](mob.html#proc/has_movespeed_modifier)
+* [update_movespeed](mob.html#proc/update_movespeed)
+*/
+
+//ANY ADD/REMOVE DONE IN UPDATE_MOVESPEED MUST HAVE THE UPDATE ARGUMENT SET AS FALSE!
+
+GLOBAL_LIST_EMPTY(movespeed_modification_cache)
+
+/// Grabs a STATIC MODIFIER datum from cache. YOU MUST NEVER EDIT THESE DATUMS, OR IT WILL AFFECT ANYTHING ELSE USING IT TOO!
+/proc/get_cached_movespeed_modifier(modtype)
+	if(!ispath(modtype, /datum/movespeed_modifier))
+		CRASH("[modtype] is not a movespeed modification typepath.")
+	var/datum/movespeed_modifier/M = modtype
+	if(initial(M.variable))
+		CRASH("[modtype] is a variable modifier, and can never be cached.")
+	M = GLOB.movespeed_modification_cache[modtype]
+	if(!M)
+		M = GLOB.movespeed_modification_cache[modtype] = new modtype
+	return M
+
+///Add a move speed modifier to a mob. If a variable subtype is passed in as the first argument, it will make a new datum. If ID conflicts, it will overwrite the old ID.
+/mob/proc/add_movespeed_modifier(datum/movespeed_modifier/type_or_datum, update = TRUE)
+	if(ispath(type_or_datum))
+		if(!initial(type_or_datum.variable))
+			type_or_datum = get_cached_movespeed_modifier(type_or_datum)
+		else
+			type_or_datum = new type_or_datum
+	var/datum/movespeed_modifier/existing = LAZYACCESS(movespeed_modification, type_or_datum.id)
+	if(existing)
+		if(existing == type_or_datum)		//same thing don't need to touch
+			return TRUE
+		remove_movespeed_modifier(existing, FALSE)
+	if(length(movespeed_modification))
+		BINARY_INSERT(type_or_datum.id, movespeed_modification, datum/movespeed_modifier, type_or_datum, priority, COMPARE_VALUE)
+	LAZYSET(movespeed_modification, type_or_datum.id, type_or_datum)
+	if(update)
+		update_movespeed()
+	return TRUE
+
+/// Remove a move speed modifier from a mob, whether static or variable.
+/mob/proc/remove_movespeed_modifier(datum/movespeed_modifier/type_id_datum, update = TRUE)
+	if(ispath(type_id_datum))
+		type_id_datum = initial(type_id_datum.id)
+	else if(!istext(type_id_datum))		//if it isn't text it has to be a datum, as it isn't a type.
+		type_id_datum = type_id_datum.id
+	if(!LAZYACCESS(movespeed_modification, type_id_datum))
+		return FALSE
+	LAZYREMOVE(movespeed_modification, type_id_datum)
+	if(update)
+		update_movespeed(FALSE)
+	return TRUE
+
+/// Used for variable slowdowns like hunger/health loss/etc, works somewhat like the old list-based modification adds. Returns the modifier datum if successful
+/mob/proc/add_or_update_variable_movespeed_modifier(datum/movespeed_modifier/type_id_datum, update = TRUE, multiplicative_slowdown)
+	/*
+	How this SHOULD work is:
```

Comment:
Why isn't this in the autodoc comment for the proc

Reply:
fixed